diff --git a/ports/esp32/main.c b/ports/esp32/main.c
index 4f0c27ee0..7bef35416 100644
--- a/ports/esp32/main.c
+++ b/ports/esp32/main.c
@@ -152,6 +152,13 @@ soft_reset:
         }
     }
 
+    // Check authentication and PIN before getting REPL
+    ret = pyexec_frozen_module("modules/_auth.py", false);
+    if (ret & PYEXEC_FORCED_EXIT) {
+       // Authentication failed, perform soft reboot
+       goto soft_reset_exit;
+    }
+
     for (;;) {
         if (pyexec_mode_kind == PYEXEC_MODE_RAW_REPL) {
             vprintf_like_t vprintf_log = esp_log_set_vprintf(vprintf_null);
